core/authn.go

```diff
@@ -34,26 +36,54 @@ type tokenResult struct {
 
 func (a *apiAuthn) handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		err := a.auth(req)
-		if err != nil {
-			errorFormatter.Write(req.Context(), rw, err)
+		token, err0 := a.tokenAuth(req)
+		ctx := req.Context()
+		if err0 == nil && token != "" {
+			// if this request was successfully authenticated, pass the token along
+			ctx = context.WithValue(ctx, "token", token)
+		}
+
+		err1 := a.localhostAuth(req)
+		if err1 == nil {
+			ctx = context.WithValue(ctx, "localhost", true)
+		}
+
+		// TODO(tessr): move this to authz as part of ACL work
+		if err0 != nil {
+			errorFormatter.Write(ctx, rw, err0)
 			return
 		}
-		next.ServeHTTP(rw, req)
+
+		next.ServeHTTP(rw, req.WithContext(ctx))
 	})
 }
 
-func (a *apiAuthn) auth(req *http.Request) error {
+func (a *apiAuthn) localhostAuth(req *http.Request) error {
+	h, _, err := net.SplitHostPort(req.RemoteAddr)
+	if err != nil {
+		return errors.Wrap(err)
+	}
+	if !net.ParseIP(h).IsLoopback() {
+		return errNotAuthenticated
+	}
+	return nil
+}
+
+func (a *apiAuthn) tokenAuth(req *http.Request) (string, error) {
 	user, pw, ok := req.BasicAuth()
+	// TODO(tessr): remove the following clause once ACLs are in place
 	if !ok && a.alt != nil && a.alt(req) {
-		return nil
+		return "", nil
 	}
 
+	// Is this the way we want to encode the token to be passed around? Or just user?
```

Comment:
Just user. I think a reasonable conceptual framework is that the "credentials" prove your "identity", and the authn handler checks the credentials but everything else works just in terms of identity, relying on authn to have done its thing.

So an access token e.g. `kr:abc123` would prove my identity as `user("kr")`.

Or a remote addr of `127.0.0.1:12345` would prove my identity as `localhost(true)`.